### PR TITLE
Add CompileOptions.options to pass extra options to JavaCompiler

### PR DESCRIPTION
**Pull request by devinrsmith**
I'd like to add the ability to add extra arguments to the compiler (only prototyped for Java 8 ATM). I'm not sure the easiest way to actually test the extra options in context of unit tests. Would a feature like this be welcome?

### CODE REVIEW

**Comment by lukaseder** [comment]
Thanks for your suggestion. I agree, extra options would be useful.

Two issues:

1. When the extra options contain `-classpath`, then jOOR shouldn't add the classpath
2. I'd prefer to call this list "options" as that's what `JavaCompiler.getTask()` calls the parameter

Will fix both of the above.

File: jOOR-java-8/src/main/java/org/joor/CompileOptions.java [hunk 1]
```diff
@@ -28,25 +28,41 @@
...
+        return new CompileOptions(newProcessors, Collections.emptyList());
+    }
+
+    public final CompileOptions extraOptions(String... extraOptions) {
+        return extraOptions(Arrays.asList(extraOptions));
+    }
+
+    public final CompileOptions extraOptions(List<String> extraOptions) {
+        return new CompileOptions(Collections.emptyList(), extraOptions);
+    }
```
Comment (lukaseder) [inline jOOR-java-8/src/main/java/org/joor/CompileOptions.java]: This is an immutable setter, so it should not set an empty processors list

File: jOOR-java-8/src/main/java/org/joor/CompileOptions.java [hunk 2]
```diff
@@ -28,25 +28,41 @@
...
     }
 
     public final CompileOptions processors(Processor... newProcessors) {
         return processors(Arrays.asList(newProcessors));
     }
 
     public final CompileOptions processors(List<? extends Processor> newProcessors) {
-        return new CompileOptions(newProcessors);
+        return new CompileOptions(newProcessors, Collections.emptyList());
+    }
```
Comment (lukaseder) [inline jOOR-java-8/src/main/java/org/joor/CompileOptions.java]: This is an immutable setter, so it should not set an empty extraOptions list

**Comment by lukaseder** [comment]
> only prototyped for Java 8 ATM

The main code base is the one not targeted at any JDK. The JDK 8 code base is copied from the main code base by a code generator. Just for the next time, please edit the main code base only

File: jOOR-java-8/src/main/java/org/joor/CompileOptions.java [hunk 3]
```diff
@@ -28,25 +28,41 @@
...
+        return extraOptions(Arrays.asList(extraOptions));
+    }
+
+    public final CompileOptions extraOptions(List<String> extraOptions) {
+        return new CompileOptions(Collections.emptyList(), extraOptions);
+    }
+
+    public final CompileOptions compileOptions(List<? extends Processor> newProcessors, List<String> extraOptions) {
+        return new CompileOptions(newProcessors, extraOptions);
     }
```
Comment (lukaseder) [inline jOOR-java-8/src/main/java/org/joor/CompileOptions.java]: There's no point for this setter that overrides both properties...

File: jOOR-java-8/src/test/java/org/joor/test/CompileOptionsTest.java [hunk 4]
```diff
@@ -77,6 +77,27 @@ public void testCompileWithAnnotationProcessors() throws Exception {
...
+        // todo: check .class version, or create better test for extraOptions
+
+        Class<?> source8Class = Reflect.compile(
+            "org.joor.test.Source8",
+            "package org.joor.test; "
+                + "public class Source8 {"
+                + "}",
+            new CompileOptions().extraOptions("-source", "8")).create().get().getClass();
+
+        // todo: check .class version, or create better test for extraOptions
```
Comment (lukaseder) [inline jOOR-java-8/src/test/java/org/joor/test/CompileOptionsTest.java]: We can implement a test that uses syntax that wasn't yet available in a given Java version. For example, default methods in Java 7, or var in Java 8

**Comment by lukaseder** [comment]
Alright, can confirm this works very well, with the improved tests that also test against working/failing source versions.

Thanks again for the contribution!

**Comment by devinrsmith** [comment]
Ha, thanks for the feedback and quickly fixing the PR yourself!

Cheers.

**Pull request merged**